Firebase/Auth/Source/FIRAuthURLPresenter.m

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRAuthURLPresenter.h"
+
+#import "FIRAuthErrorUtils.h"
+#import "FIRAuthUIDelegate.h"
+
+@import SafariServices;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FIRAuthDefaultUIDelegate : NSObject <FIRAuthUIDelegate>
+/** @fn defaultUIDelegate
+    @brief Returns a default FIRAuthUIDelegate object.
+    @return The default FIRAuthUIDelegate object.
+ */
++ (id<FIRAuthUIDelegate>)defaultUIDelegate;
+@end
+
+@interface FIRAuthURLPresenter () <SFSafariViewControllerDelegate>
+@end
+
+@implementation FIRAuthURLPresenter {
+  /** @var _callbackMatcher
+      @brief The callback URL matcher for the current presentation, if one is active.
+   */
+  FIRAuthURLCallbackMatcher _Nullable _callbackMatcher;
+
+  /** @var _safariViewController
+      @brief The SFSafariViewController used for the current presentation, if any.
+   */
+  SFSafariViewController *_Nullable _safariViewController;
+
+  /** @var _completion
+      @brief The completion handler for the current presentaion, if one is active.
+      @remarks This variable is also used as a flag to indicate a presentation is active.
+   */
+  FIRAuthURLPresentationCompletion _Nullable _completion;
+}
+
+- (void)presentURL:(NSURL *)URL
+        UIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
+   callbackMatcher:(FIRAuthURLCallbackMatcher)callbackMatcher
+        completion:(FIRAuthURLPresentationCompletion)completion {
+  _callbackMatcher = callbackMatcher;
+  _completion = completion;
+  // If a UIDelegate is not provided.
+  if (!UIDelegate) {
+    id<FIRAuthUIDelegate> delegate = [FIRAuthDefaultUIDelegate defaultUIDelegate];
+    [self presentWebContextWithController:delegate URL:URL];
+    return;
+  }
+  // If a valid UIDelegate is provided.
+  [self presentWebContextWithController:UIDelegate URL:URL];
+}
+
+- (BOOL)canHandleURL:(NSURL *)URL {
+  if (_callbackMatcher(URL)) {
+    _callbackMatcher = nil;
+    [self finishPresentationWithURL:URL error:nil];
+    return YES;
+  }
+  return NO;
+}
+
+/** @fn presentWebContextWithController:URL:
+    @brief Presents a SFSafariViewController or WKWebView to display the contents of the URL
+        provided.
+    @param controller The controller used to present the SFSafariViewController or WKWebView.
+    @param URL The URL to display in the SFSafariViewController or WKWebView.
+ */
+- (void)presentWebContextWithController:(id)controller URL:(NSURL *)URL {
+#if HAS_SAFARI_VIEW_CONTROLLER
+   if (_safariViewController) {
+    // Unable to start a new presentation on top of modal SFSVC presentation.
+    _completion(nil, [FIRAuthErrorUtils webContextAlreadyPresentedErrorWithMessage:nil]);
+    return;
+  }
+  SFSafariViewController *safariViewController = [[SFSafariViewController alloc] initWithURL:URL];
```

Comment:
While it is fine that this won't work on iOS 8-, please make sure tests don't fail on iOS 8-.

Reply:
Done, had to update unit tests.